src/CodeGen/ODataT4CodeGenTemplate.ttinclude

```diff
@@ -0,0 +1,1333 @@
+    internal abstract void WriteCastToMethods(string baseTypeName, string derivedTypeName, string derivedTypeFullName, string returnTypeName);
+    internal abstract void WriteBoundFunctionReturnSingleResultAsExtension(string functionName, string originalFunctionName, string boundTypeName, string returnTypeName, string returnTypeNameWithSingleSuffix, string parameters, string fullNamespace, string parameterValues, bool isComposable, bool isReturnEntity, bool useEntityReference);
+    internal abstract void WriteBoundFunctionReturnCollectionResultAsExtension(string functionName, string originalFunctionName, string boundTypeName, string returnTypeName, string parameters, string fullNamespace, string parameterValues, bool isComposable, bool useEntityReference);
+    internal abstract void WriteBoundActionAsExtension(string actionName, string originalActionName, string boundSourceType, string returnTypeName, string parameters, string fullNamespace, string parameterValues);
+    #endregion Language specific write methods.
+
+    internal HashSet<EdmPrimitiveTypeKind> ClrReferenceTypes { get { 
+        if (clrReferenceTypes == null)
+        {
+            clrReferenceTypes = new HashSet<EdmPrimitiveTypeKind>()
+            {
+                EdmPrimitiveTypeKind.String, EdmPrimitiveTypeKind.Binary, EdmPrimitiveTypeKind.Geography, EdmPrimitiveTypeKind.Stream,
+                EdmPrimitiveTypeKind.GeographyPoint, EdmPrimitiveTypeKind.GeographyLineString, EdmPrimitiveTypeKind.GeographyPolygon,
+                EdmPrimitiveTypeKind.GeographyCollection, EdmPrimitiveTypeKind.GeographyMultiPolygon, EdmPrimitiveTypeKind.GeographyMultiLineString,
+                EdmPrimitiveTypeKind.GeographyMultiPoint, EdmPrimitiveTypeKind.Geometry, EdmPrimitiveTypeKind.GeometryPoint,
+                EdmPrimitiveTypeKind.GeometryLineString, EdmPrimitiveTypeKind.GeometryPolygon, EdmPrimitiveTypeKind.GeometryCollection,
+                EdmPrimitiveTypeKind.GeometryMultiPolygon, EdmPrimitiveTypeKind.GeometryMultiLineString, EdmPrimitiveTypeKind.GeometryMultiPoint
+            };
+        }
+        return clrReferenceTypes;
+    } }
+    private HashSet<EdmPrimitiveTypeKind> clrReferenceTypes;
+
+    /// <summary>
+    /// Generates code for the OData client.
+    /// </summary>
+    /// <returns>The generated code for the OData client.</returns>
+    public override string TransformText()
+    {
+        context.MultipleFilesManager.StartHeader();
+        this.WriteFileHeader();
+        context.MultipleFilesManager.EndBlock();
+        this.WriteNamespaces();
+        context.MultipleFilesManager.Process(context.SplitGeneratedFileIntoMultipleFiles);
+        return context.MultipleFilesManager.template.ToString();
+    }
+
+    internal void WriteNamespaces()
+    {
+        foreach(string fullNamespace in context.NamespacesInModel)
+        {
+            this.WriteNamespace(fullNamespace);
+        }
+    }
+
+    internal void WriteNamespace(string fullNamespace)

```

Comment:
This function is really big, and I think it can be broken down to smaller functions.

Reply:
Fixed